pyasice/ocsp.py

```diff
@@ -0,0 +1,222 @@
+from oscrypto import asymmetric
+
+from .exceptions import PyAsiceError
+from .signature_verifier import verify
+
+
+class SKHackedTBSRequestExtension(TBSRequestExtension):
+    """A workaround class for compatibility with old java libraries used in SK.ee
+
+    The jdigidoc library used in the SK service fails to validate an OCSP request
+    made up with the original TBSRequestExtension because it expects a plain OctetString in the value field.
+    This does not break validation with newer libraries including digidoc-tool
+    """
+
+    _fields = [
+        ("extn_id", TBSRequestExtensionId),
+        ("critical", Boolean, {"default": False}),
+        ("extn_value", OctetString),  # This replaces ParsableOctetString
+    ]
+
+
+class OCSPError(PyAsiceError):
+    pass
+
+
+class OCSP(object):
+    """
+    Certificate validation request via the OCSP protocol, using the asn1crypto/ocspbuilder stack.
+
+    Perform certificate validation:
+
+        ocsp = OCSP(ocsp_service_url)
+        ocsp.validate(subject_cert, issuer_cert, signature)
+
+    Check an arbitrary response:
+
+        ocsp = OCSP.load(der_encoded_binary_data)
+    """
+
+    REQUEST_CONTENT_TYPE = "application/ocsp-request"
+    RESPONSE_CONTENT_TYPE = "application/ocsp-response"
+
+    def __init__(self, url=None):
+        """"""
+        self.url = url
+        self.ocsp_response: Optional[OCSPResponse] = None
+
+    def validate(self, subject_cert, issuer_cert, signature):
+        """
+        A correct OCSP request for BDOC-TM shall contain a DER-encoded "Hash value of the binary value of
+        the signature along with hash algorithm identifier" in the `nonce` field. Yet at the moment, apparently
+        the OCSP service doesn't qualify for time-mark BDOCs so this is probably not significant
+
+        Provide certificates as asn1crypto.x509.Certificate object, oscrypto.asymmetric.Certificate object,
+          a buffer (bytes) in PEM or DER format, or a path to a certificate file.
+
+        :param issuer_cert: a certificate object, a buffer or a path to file
+        :param subject_cert: a certificate object, a buffer or a path to file
+        :param bytes signature: required for BDOC-TM format
+        :return:
+        """
+        ocsp_request = self.build_ocsp_request(subject_cert, issuer_cert, signature)
+
+        try:
+            response = requests.post(
+                self.url,
+                data=ocsp_request.dump(),
+                headers={"Content-Type": self.REQUEST_CONTENT_TYPE, "Connection": "close"},
+            )
+            response.raise_for_status()
+        except requests.ConnectionError:
+            raise OCSPError("Failed to connect to OCSP service at {}".format(self.url))
+        except requests.HTTPError as e:
+            raise OCSPError("Bad response from OCSP service at {}: {}".format(self.url, e))
+
+        assert response.status_code == 200  # this wouldn't tell anything yet
```

Comment:
Don't like the usage of asserts for runtime issues. Asserts will be disabled when running in optimized mode and that could unexpectedly disable the safeguards we have.

Reply:
Thanks, valid remark, got rid of asserts.